raiden/network/proxies/token_network.py

```diff
@@ -812,96 +846,168 @@ def _set_total_deposit(
-            if transaction_executed:
-                block = receipt_or_none["blockNumber"]
-            else:
-                block = checking_block
+            if receipt:
+                # Because the gas estimation succeeded it is known that:
+                # - The channel was open.
+                # - The account had enough tokens to deposit
+                # - The account had enough balance to pay for the gas (however
+                #   there is a race condition for multiple transactions #3890)
+                failed_at_blockhash = encode_hex(receipt["blockHash"])
+
+                if receipt["cumulativeGasUsed"] == gas_limit:
+                    msg = (
+                        f"setTotalDeposit failed and all gas was used "
+                        f"({gas_limit}). Estimate gas may have underestimated "
+                        f"setTotalDeposit, or succeeded even though an assert is "
+                        f"triggered, or the smart contract code has an "
+                        f"conditional assert."
+                    )
+                    raise RaidenUnrecoverableError(msg)
+
+                # Query the current state to check for transaction races
+                our_details = self._detail_participant(
+                    channel_identifier=channel_identifier,
+                    detail_for=self.node_address,
+                    partner=partner,
+                    block_identifier=failed_at_blockhash,
+                )
+                partner_details = self._detail_participant(
+                    channel_identifier=channel_identifier,
+                    detail_for=self.node_address,
+                    partner=partner,
+                    block_identifier=failed_at_blockhash,
+                )
+
+                channel_state = self._get_channel_state(
+                    participant1=self.node_address,
+                    participant2=partner,
+                    block_identifier=failed_at_blockhash,
+                    channel_identifier=channel_identifier,
+                )
+                # Check if deposit is being made on a nonexistent channel
+                if channel_state != ChannelState.OPENED:
+                    msg = "Deposit failed as the channel was not open"
+                    raise RaidenUnrecoverableError(msg)
+
+                deposit_amount = total_deposit - our_details.deposit
+
+                total_channel_deposit = total_deposit + partner_details.deposit
+                if total_channel_deposit >= UINT256_MAX:
+                    raise RaidenUnrecoverableError("Deposit overflow")
+
+                total_deposit_done = our_details.deposit >= total_deposit
+                if total_deposit_done:
+                    raise DepositMismatch("Requested total deposit was already performed")
+
+                channel_participant_deposit_limit = self.channel_participant_deposit_limit
+                if total_deposit > channel_participant_deposit_limit:
```

Comment:
Please move the this check to the end, it is possible that the transaction failed because of the token network limit, and that one is recoverable.

Reply:
Additionally, this check would only fail if the `channel_participant_deposit_limit` changed. And to properly check this, it should not be a property but a call with the correct block number.